modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMapper.java

```diff
@@ -0,0 +1,563 @@
+import org.apache.ignite.internal.sql.engine.rel.IgniteLimit;
+import org.apache.ignite.internal.sql.engine.rel.IgniteMergeJoin;
+import org.apache.ignite.internal.sql.engine.rel.IgniteNestedLoopJoin;
+import org.apache.ignite.internal.sql.engine.rel.IgniteProject;
+import org.apache.ignite.internal.sql.engine.rel.IgniteReceiver;
+import org.apache.ignite.internal.sql.engine.rel.IgniteRel;
+import org.apache.ignite.internal.sql.engine.rel.IgniteRelVisitor;
+import org.apache.ignite.internal.sql.engine.rel.IgniteSender;
+import org.apache.ignite.internal.sql.engine.rel.IgniteSort;
+import org.apache.ignite.internal.sql.engine.rel.IgniteSortedIndexSpool;
+import org.apache.ignite.internal.sql.engine.rel.IgniteSystemViewScan;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTableFunctionScan;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTableModify;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTableScan;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTableSpool;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTrimExchange;
+import org.apache.ignite.internal.sql.engine.rel.IgniteUnionAll;
+import org.apache.ignite.internal.sql.engine.rel.IgniteValues;
+import org.apache.ignite.internal.sql.engine.rel.ProjectableFilterableTableScan;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteColocatedHashAggregate;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteColocatedSortAggregate;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteMapHashAggregate;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteMapSortAggregate;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteReduceHashAggregate;
+import org.apache.ignite.internal.sql.engine.rel.agg.IgniteReduceSortAggregate;
+import org.apache.ignite.internal.sql.engine.rel.set.IgniteSetOp;
+import org.apache.ignite.internal.sql.engine.schema.IgniteTable;
+import org.apache.ignite.internal.sql.engine.trait.IgniteDistributions;
+import org.apache.ignite.internal.sql.engine.trait.TraitUtils;
+
+/**
+ * A mapper that traverse a fragment tree and calculates a mapping.
+ */
+class FragmentMapper {
+
+    private final RelMetadataQuery mq;
+
+    private final MappingContext context;
+
+    private final Int2ObjectMap<ExecutionTarget> targets;
+
+    FragmentMapper(RelMetadataQuery mq, MappingContext context, Int2ObjectMap<ExecutionTarget> targets) {
+        assert mq instanceof RelMetadataQueryEx;
+
+        this.mq = mq;
+        this.context = context;
+        this.targets = targets;
+    }
+
+    /**
+     * Computes mapping for the given fragment.
+     *
+     * @param fragment A fragment to compute mapping for.
+     * @return Fragment meta information.
+     */
+    public FragmentMapping map(Fragment fragment) throws FragmentMappingException {
+        Mapping mapping = fragment.root().accept(new MapperVisitor());
+
+        if (fragment.rootFragment()) {
```

Comment:
Both `if` branches looks very similar. The difference is we throws an exception if root fragment can't be mapped to the local node.

Will be the next statement valid: `fragment.rootFragment()==true` only if `fragment.single()==true` ?
If so, then we can simplify this and drop `else if` branch.
```suggestion
        if (fragment.single()) {
            .....
            } catch (ColocationMappingException e) {
                if (fragment.rootFragment() {
                    throw new FragmentMappingException(e.getMessage(), fragment.root(), e);
                }
                
                ....
             }
        }
```

Reply:
> Will be the next statement valid: fragment.rootFragment()==true only if fragment.single()==true ?

currently, yes, that's correct. So, I've simplified this method as you suggest